<!DOCTYPE HTML>
<html>
 <head>
  <meta charset="utf-8"/>
  <title>
   Made with Remarkable!
  </title>
  <link href="http://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.1/styles/github.min.css" rel="stylesheet"/>
  <style type="text/css">
   body,table tr{background-color:#fff}table tr td,table tr th{border:1px solid #ccc;text-align:left;padding:6px 13px;margin:0}pre code,table,table tr{padding:0}hr,pre code{background:0 0}body{font:16px Helvetica,Arial,sans-serif;line-height:1.4;color:#333;word-wrap:break-word;padding:10px 15px}strong,table tr th{font-weight:700}h1{font-size:2em;margin:.67em 0;text-align:center}h2{font-size:1.75em}h3{font-size:1.5em}h4{font-size:1.25em}h1,h2,h3,h4,h5,h6{font-weight:700;position:relative;margin-top:15px;margin-bottom:15px;line-height:1.1}h1,h2{border-bottom:1px solid #eee}hr{height:0;margin:15px 0;overflow:hidden;border:0;border-bottom:1px solid #ddd}a{color:#4183C4}a.absent{color:#c00}ol,ul{padding-left:15px;margin-left:5px}ol{list-style-type:lower-roman}table tr{border-top:1px solid #ccc;margin:0}table tr:nth-child(2n){background-color:#aaa}table tr td :first-child,table tr th :first-child{margin-top:0}table tr td:last-child,table tr th :last-child{margin-bottom:0}img{max-width:100%}blockquote{padding:0 15px;border-left:4px solid #ccc}code,tt{margin:0 2px;padding:0 5px;white-space:nowrap;border:1px solid #eaeaea;background-color:#f8f8f8;border-radius:3px}pre code{margin:0;white-space:pre;border:none}.highlight pre,pre{background-color:#f8f8f8;border:1px solid #ccc;font-size:13px;line-height:19px;overflow:auto;padding:6px 10px;border-radius:3px}
  </style>
 </head>
 <body>
  <blockquote>
   <p>
    <strong>
     Author
    </strong>
    : Adolfo Hoyos Díaz
    <br/>
    <strong>
     Advisor
    </strong>
    : Diana Hermith Ph.D.
   </p>
  </blockquote>
  <h1 id="session-6-final-of-module-1">
   Session 6: Final of Module 1
  </h1>
  <p>
   This session brings a mini project described in
   <a href="project.pdf">
    this file
   </a>
   .
  </p>
  <p>
   From now on on this text, I present my answers to the questionary delivered by Dr. Diana Hermith.
  </p>
  <h2 id="1-exact-vs-approx">
   1. Exact vs Approx.
  </h2>
  <p>
   In this part of the practice, we study two dynamics described in
   <a href="1/MMexact.bngl">
    exact dynamics file
   </a>
   and the [approximate dynamics file][1/approx_source].
  </p>
  <p>
   The
   <strong>
    exact
   </strong>
   dynamics has the next reactions
  </p>
  <blockquote>
   <p>
    <strong>
     ES bind
    </strong>
    reaction:
   </p>
   <p>
    <mathjax>
     $E + S \longleftrightarrow ES$
    </mathjax>
   </p>
   <p>
    detailed as:
    <br/>
    <mathjax>
     $E(s) + S(Y \sim 0) \array{k_{p_1} \\   \longleftrightarrow \\ k_{m_1} } E(s!1).S(Y  \sim 0!1)$
    </mathjax>
    <br/>
    with
   </p>
   <p>
    <mathjax>
     $k_{p_1} = \frac {1} {N_{Av}}$
    </mathjax>
    <br/>
    <mathjax>
     $k_{m_1}=\frac{1}{10}$
    </mathjax>
   </p>
   <p>
    <strong>
     ESconvert
    </strong>
    :
   </p>
   <p>
    <mathjax>
     $ES \rightarrow E+S$
    </mathjax>
   </p>
   <p>
    detailed as:
    <br/>
    <mathjax>
     $\array{ \\E(s!1).S(Y \sim 0!1)} \array{k_2 \\ \longrightarrow} \array{ \\E(s) + S(Y \sim P) }$
    </mathjax>
   </p>
   <p>
    with
   </p>
   <p>
    <mathjax>
     $k_{2}= \frac{1}{100}$
    </mathjax>
   </p>
  </blockquote>
  <p>
   With the initial concentrations
  </p>
  <blockquote>
   <p>
    <mathjax>
     $ E_0 =\frac{1}{100}\times N_{Av}$
    </mathjax>
    <br/>
    <mathjax>
     $S_0 = 1\times N_{Av}$
    </mathjax>
   </p>
  </blockquote>
  <p>
   Meanwhile the
   <strong>
    approx
   </strong>
   file describes only one reaction:
  </p>
  <blockquote>
   <p>
    <mathjax>
     $E+S\rightarrow E+S$
    </mathjax>
   </p>
   <p>
    detailed as:
    <br/>
    <mathjax>
     $\array{ \\ E(s) + S(Y \sim 0)} \array{\frac{kcat}{(Km + Su)} \\ \longrightarrow} \array{ \\E(s) + S(Y \sim P)  }$
    </mathjax>
   </p>
   <p>
    with
   </p>
   <p>
    <mathjax>
     $k_f=\frac{1}{N_{Av}}$
    </mathjax>
    <br/>
    <mathjax>
     $k_r=\frac{1}{10}$
    </mathjax>
    <br/>
    <mathjax>
     $k_{cat}=\frac{1}{100}$
    </mathjax>
    <br/>
    <mathjax>
     $k_m=\frac{k_{cat}+k_r}{k_f}$
    </mathjax>
   </p>
  </blockquote>
  <p>
   with the same initial concentrations.
  </p>
  <p>
   We can see that both reactions start with the same concentrations of enzyme and
   <strong>
    dephosporylated
   </strong>
   substrate and, although the
   <em>
    exact reaction
   </em>
   describes an intermediate step in which the initial seed converts into the
   <strong>
    ES
   </strong>
   substance, both reactions end with enzyme and phosphorylated substrate. We are yet to check, through simulation, whether the result is the same or not… most probably not:
  </p>
  <p>
   The reaction rate of the
   <strong>
    approx
   </strong>
   file
  </p>
  <p>
   <mathjax>
    $MM_{rate} =\frac{kcat}{(k_m + Su)}$
   </mathjax>
  </p>
  <p>
   depends on the concentration of dephosporylated compound SU; this will speed up the reaction since the rate will grow as the concencentration of SU decreases.
  </p>
  <table>
   <thead>
    <tr>
     <th>
      Feature
     </th>
     <th>
      Units
     </th>
     <th>
      Exact
     </th>
     <th>
      Aprox
     </th>
    </tr>
   </thead>
   <tbody>
    <tr>
     <td>
      Volume
     </td>
     <td>
      <mathjax>
       ${um}^3$
      </mathjax>
     </td>
     <td>
      1
     </td>
     <td>
      1
     </td>
    </tr>
   </tbody>
  </table>
  <h3 id="simulation-results">
   Simulation results
  </h3>
  <h4 id="exact">
   Exact
  </h4>
  <p>
   <img alt="exact" src="1/exact.png"/>
  </p>
  <p>
   Stability of the process is reached around
   <mathjax>
    $1.4\times 10 ^{4} s$
   </mathjax>
   , and also we can see that the transition state is very ephimeral, so the concentration of ES compound remains very low during the whole process, as shown on the next plot:
  </p>
  <p>
   <img alt="exactES" src="1/exactES.png"/>
  </p>
  <p>
   ES compund starts from
   <mathjax>
    $\sim5.5$
   </mathjax>
   and goes to
   <mathjax>
    $0$
   </mathjax>
   .
  </p>
  <h4 id="approx">
   Approx
  </h4>
  <p>
   despite it may not ve obvious, the next graphic, obtained simulating the
   <strong>
    approx
   </strong>
   file, (almost) exactly overlaps with the previous graphic.
  </p>
  <p>
   <img alt="approx" src="1/approx.png"/>
  </p>
  <p>
   We apreciate that the ES compound is being monitored as an observable variable, even when this substance is not present in the described reaction. This leads to the next graphic:
  </p>
  <p>
   <img alt="approxES" src="1/approxES.png"/>
  </p>
  <p>
   that shows this variable always as zero, as should be expected.
  </p>
  <p>
   The results for the
   <mathjax>
    $MM_{rate}$
   </mathjax>
   function is seen in the next figure
  </p>
  <p>
   <img alt="approxMMrate" src="1/approxMMrate.png"/>
  </p>
  <p>
   and, although the wave form relates (almost) exactly to the ES evolution of the
   <strong>
    exact
   </strong>
   reaction, the values are two orders of magnitud below, going from
   <mathjax>
    $\sim 5.5 \times 10 ^{-2}$
   </mathjax>
   to
   <mathjax>
    $0$
   </mathjax>
   .
  </p>
  <h2 id="2-a-little-modification">
   2. A little modification
  </h2>
  <p>
   For this part, we added the function:
  </p>
  <blockquote>
   <p>
    <mathjax>
     $ES_{eff} = E_0 \times \frac{Su}{k_m + S_u}$
    </mathjax>
   </p>
  </blockquote>
  <p>
   let’s remember we have
  </p>
  <blockquote>
   <p>
    <mathjax>
     $k_f=\frac{1}{N_{Av}}$
    </mathjax>
    <br/>
    <mathjax>
     $k_r=\frac{1}{10}$
    </mathjax>
    <br/>
    <mathjax>
     $k_{cat}=\frac{1}{100}$
    </mathjax>
    <br/>
    <mathjax>
     $k_m=\frac{k_{cat}+k_r}{k_f}$
    </mathjax>
   </p>
  </blockquote>
  <p>
   so
  </p>
  <blockquote>
   <p>
    <mathjax>
     $ES_{eff} = E_0 \times \frac{Su}{\frac{k_{cat}+k_r}{k_f} + Su}$
    </mathjax>
    <br/>
    The graph obtained for this function from the simulation results is shown next, and it overlaps with the ES complex of the
    <strong>
     exact
    </strong>
    reaction.
   </p>
  </blockquote>
  <p>
   <img alt="ESeff" src="2/approxESeff.png"/>
  </p>
  <p>
   <img alt="ESMM" src="2/approxESMM.png"/>
  </p>
  <h2 id="3-brusselator-model">
   3. Brusselator model
  </h2>
  <p>
   The proposed reaction scheme is described with the following equations:
  </p>
  <p>
   <mathjax>
    $R_1:\array{ &amp;   k_1 \\  A &amp; \longrightarrow &amp; X} $
   </mathjax>
  </p>
  <p>
   <mathjax>
    $R_2: \array{ &amp; k_2 \\ B+X &amp; \longrightarrow &amp; Y+D}$
   </mathjax>
  </p>
  <p>
   <mathjax>
    $R_3: \array{ &amp; k_3 \\ 2X+Y &amp; \longrightarrow &amp; 3X}$
   </mathjax>
  </p>
  <p>
   <mathjax>
    $R_4: \array{ &amp; k_4 \\ X &amp; \longrightarrow &amp; C}$
   </mathjax>
  </p>
  <p>
   Only by taking a quick look at these equation, we can expect a complete depletion of
   <strong>
    <em>
     A
    </em>
   </strong>
   in order to generate
   <strong>
    <em>
     X
    </em>
   </strong>
   , and of
   <strong>
    <em>
     X
    </em>
   </strong>
   to generate
   <strong>
    <em>
     C
    </em>
   </strong>
   . Also, we should expect some byproducts of this main reaction, like some remaining
   <strong>
    <em>
     B
    </em>
   </strong>
   ,
   <strong>
    <em>
     Y
    </em>
   </strong>
   and
   <strong>
    <em>
     D
    </em>
   </strong>
   ; this of course depends on the initial concentrations of all components.
  </p>
  <p>
   A
   <em>
    Biocham
   </em>
   code that describes these equations, with only some concentrrations of
   <strong>
    <em>
     A
    </em>
   </strong>
   and
   <strong>
    <em>
     B
    </em>
   </strong>
   are considered initially, is shown next:
  </p>
  <pre><code class="python">MA (K1) for A =&gt; X.
MA (K2) for B + X =&gt; Y + D.
MA (K3) for 2*X + Y =&gt; 3*X.
MA (K4) for X =&gt; C.
parameter(K1, 1).
parameter(K2, 1).
parameter(K3, 1).
parameter(K4, 1).
present(A, 1).
present(B, 1).
</code></pre>
  <p>
   These initial concentrations were selected as shown in the previous code since component
   <strong>
    <em>
     A
    </em>
   </strong>
   and
   <strong>
    <em>
     B
    </em>
   </strong>
   would generate the rest of the necessary substances of the reaction scheme.
  </p>
  <hr/>
  <h3 id="simulations">
   Simulations
  </h3>
  <p>
   The ordinary differencial equations
   <strong>
    (ODE) simulation
   </strong>
   results are shown on next figure:
  </p>
  <p>
   <img alt="" src="3/images/ODE.PNG"/>
  </p>
  <p>
   As expected before, we can observe that
   <strong>
    <em>
     A
    </em>
   </strong>
   is completely consumed. Also, although there is no
   <strong>
    <em>
     X
    </em>
   </strong>
   ‘s initial concentration, it is initially produced (from A and then with help of
   <strong>
    <em>
     Y
    </em>
   </strong>
   ), but later is completely consumed. We also can see that the initial substance
   <strong>
    <em>
     B
    </em>
   </strong>
   does not completely react and a little more than half of it (~55%) still remains at stabilization time. At this same time, the most abundant substance is
   <strong>
    <em>
     C
    </em>
   </strong>
   , and the byproducts
   <strong>
    <em>
     D
    </em>
   </strong>
   and
   <strong>
    <em>
     Y
    </em>
   </strong>
   are not completely utilized. There is less of
   <strong>
    <em>
     Y
    </em>
   </strong>
   than of
   <strong>
    <em>
     D
    </em>
   </strong>
   since the former reacts with
   <strong>
    <em>
     X
    </em>
   </strong>
   to produce more
   <strong>
    <em>
     X
    </em>
   </strong>
   .
  </p>
  <p>
   Additionally, a
   <strong>
    stochastic simulation
   </strong>
   (using Gillespie method), which describes a very similar behavior than the previous one, is shown next:
  </p>
  <p>
   ![][brusStocastic
  </p>
  <p>
   On the other hand, Biocham’s boolean simulator gave the next result:
  </p>
  <p>
   <img alt="" src="3/images/boolean.PNG"/>
  </p>
  <blockquote>
   <p>
    We need to have in mid that, we most likely won’t get the same plots from the stochastic and Boolean simulators, since they are, well, stochastic: they model a non-deterministic or random behavior.
   </p>
  </blockquote>
  <p>
   The shown Boolean simulation plot also suggests a depletion of
   <strong>
    <em>
     A
    </em>
   </strong>
   and
   <strong>
    <em>
     X
    </em>
   </strong>
   , while all other components remain in the system:
   <strong>
    <em>
     C
    </em>
   </strong>
   is generated before the simultaneouly appearing
   <strong>
    <em>
     D
    </em>
   </strong>
   and
   <strong>
    <em>
     Y
    </em>
   </strong>
   , while
   <strong>
    <em>
     B
    </em>
   </strong>
   is always present.
  </p>
  <hr/>
  <h3 id="ctl">
   CTL
  </h3>
  <p>
   In this section I’m going to show the evaluation results of the proposed CTL queries:
  </p>
  <ul>
   <li>
    Reachable
   </li>
   <li>
    Stable
   </li>
   <li>
    Steady
   </li>
  </ul>
  <h4 id="reachable">
   Reachable
  </h4>
  <p>
   This boolean query tells us if the expression we are evaluating is going to happen somewhere in time. Of course, when we submit components of this reaction, and as we have seen in the previous simulation results, all the components are present in the system at a given moment. This is seen as the result of the query being a
   <strong>
    “true”
   </strong>
   value.
  </p>
  <p>
   <img alt="" src="3/images/CTL.png"/>
  </p>
  <p>
   When we try a
   <strong>
    reachable
   </strong>
   (or any other) property query of a molecule that we have not defined in our model, we get the following error message:
  </p>
  <p>
   <img alt="" src="3/images/CTLundefined.PNG"/>
  </p>
  <h4 id="steady">
   Steady
  </h4>
  <p>
   The
   <strong>
    steady
   </strong>
   query tells us if a expression reaches a steady value in any point in time. As we have seen previously, all the types of simulations we ran have told us that all molecules reach a steady concentration in the system. This is confirmed by the boolean simulation using CTL, shown next:
  </p>
  <p>
   <img alt="" src="3/images/CTLsteady.png"/>
  </p>
  <h4 id="stable">
   Stable
  </h4>
  <p>
   The
   <strong>
    stable
   </strong>
   query tells us something like:
  </p>
  <ul>
   <li>
    <mathjax>
     $stable(X) \longrightarrow$
    </mathjax>
    I will get
    <mathjax>
     $X$
    </mathjax>
    forever no matter what.
   </li>
  </ul>
  <p>
   The results of the queries for every component of our reactions are shown next:
  </p>
  <p>
   <img alt="" src="3/images/CTLstable.png"/>
  </p>
  <p>
   Summarizing the previous results:
  </p>
  <table>
   <thead>
    <tr>
     <th>
      Component
     </th>
     <th>
      Stable?
     </th>
    </tr>
   </thead>
   <tbody>
    <tr>
     <td>
      A
     </td>
     <td>
      false
     </td>
    </tr>
    <tr>
     <td>
      B
     </td>
     <td>
      false
     </td>
    </tr>
    <tr>
     <td>
      X
     </td>
     <td>
      false
     </td>
    </tr>
    <tr>
     <td>
      Y
     </td>
     <td>
      false
     </td>
    </tr>
    <tr>
     <td>
      D
     </td>
     <td>
      true
     </td>
    </tr>
    <tr>
     <td>
      C
     </td>
     <td>
      true
     </td>
    </tr>
   </tbody>
  </table>
  <p>
   This is clear if we think that none of the reactions:
  </p>
  <p>
   <mathjax>
    $R_1:\array{ &amp;   k_1 \\  A &amp; \longrightarrow &amp; X} $
   </mathjax>
  </p>
  <p>
   <mathjax>
    $R_2: \array{ &amp; k_2 \\ B+X &amp; \longrightarrow &amp; Y+D}$
   </mathjax>
  </p>
  <p>
   <mathjax>
    $R_3: \array{ &amp; k_3 \\ 2X+Y &amp; \longrightarrow &amp; 3X}$
   </mathjax>
  </p>
  <p>
   <mathjax>
    $R_4: \array{ &amp; k_4 \\ X &amp; \longrightarrow &amp; C}$
   </mathjax>
  </p>
  <p>
   describe any consumption of
   <strong>
    <em>
     D
    </em>
   </strong>
   or
   <strong>
    <em>
     C
    </em>
   </strong>
   , so these components, once present in the system, will remain there.
  </p>
  <h2 id="4-read-and-analyze">
   4. Read and analyze
  </h2>
  <p>
   In this section, we are being asked to analyze the statement:
  </p>
  <blockquote>
   <p>
    <em>
     “Biological organisms are complex because they are characterized by hierarchies of self-assembled networks. Networks of interacting molecules at the level of the cell, interacting molecules and cells at the level of tissues, interacting organs at the level of the organisms, and interacting organisms at the level of ecosystems. These interactions/networks are well organized in space and time, normally, they are non-linear and there are a large number of them”
    </em>
   </p>
  </blockquote>
  <p>
   from the point of view of the texts:
  </p>
  <ul>
   <li>
    <a href="../essay1/TheComplexityOfCell-BiologicalSystems.pdf">
     The complexity of cell-biological systems
    </a>
   </li>
   <li>
    <a href="../essay2/Gibas2001DevelopingBioinformatics.pdf">
     Developing bioinformatics computer skills
    </a>
   </li>
  </ul>
  <p>
   So, I’m taking quotes from both texts, that argument the mentioned statement.
  </p>
  <hr/>
  <h3 id="the-complexity-of-cell-biological-systems">
   The complexity of cell-biological systems
  </h3>
  <p>
   In the
   <em>
    Introduction
   </em>
   section, page 357, the text says:
  </p>
  <blockquote>
   <ul>
    <li>
     Cells are composed of a very large number and variety of components interacting in space and time.
    </li>
    <li>
     Cell-biological systems are difficult to observe.
    </li>
    <li>
     The dynamic functioning of cells is of a nonlinear nature.
    </li>
    <li>
     Living systems are subject to continuous change.
    </li>
   </ul>
  </blockquote>
  <p>
   The first three items of the quoted list address some very important characteristics of living systems, and they summarize the main ideas of the analyzed statement, adding the non-stationary nature of life.
  </p>
  <p>
   Now, on the next page appears the next fragment:
  </p>
  <blockquote>
   <p>
    The term “self” in “self-organization” suggests a form of closure of the system to outside influences, a kind of autonomy. Physically a cell is an open system which relies on a constant exchange of energy, matter and information with its environment. The closure of a living system is therefore not with respect to material causes but with respect to efficient causation: 1 In a living system each part/process is at once cause and effect, a means and an end — the cell is a self-organizing biochemical system that fabricates itself [
    <a href="http://www.sciencedirect.com/science/article/pii/S0022519307000203)">
     Wolkenhauer, et al., 2007
    </a>
    ].
    <br/>
    Living systems are thus self-referential, every part owes its existence/explanation to the organization of the remaining parts. Through cyclic self-organizing and self-maintaining processes, parts generate the whole as does the whole define the context in which its parts function. The principle of autonomous self-fabrication is a, if not the, fundamental property that distinguishes living from non-living systems.
   </p>
  </blockquote>
  <p>
   The “self-assembled networks” mentioned by the studied statement can be related to the “self-organization” mentioned in the previous piece, and here it is discused the fact that this self-organization actually depends on all internal and external factors. This adds up to the idea of “well-organized interactions/networks, due the the explanation of their multilevel nature inside living systems.
  </p>
  <p>
   Beginning of section 7, page 377:
  </p>
  <blockquote>
   <p>
    “
    <em>
     The structural (physical/material) organization of cells is the outcome of an elaborate self-organizing process, involving gene expression, regulation, signalling and metabolism. This structural organization of the cell then serves as an environment for the cell’s functional organization, leading to growth, differentiation, proliferation and cell death (apoptosis). The cell is itself a component of a larger system with higher levels of structural and functional organization.
    </em>
    “
   </p>
  </blockquote>
  <p>
   Finally, this paragraph further explains the complexity of self-organized interactions of living systems.
  </p>
  <h3 id="developing-bioinformatics-computer-skills">
   Developing bioinformatics computer skills
  </h3>
  <p>
   From page 32, I quote:
  </p>
  <blockquote>
   <p>
    <em>
     “There is more to biology than sequences. Proteins and nucleic acids also have complex 3D structures that provide clues to their functions in the living organism. Molecular structures are usually represented as collections of atoms, each of which has a defined position in 3D space. Structure analysis can be performed on static structures, or movements and interactions in the molecules can be studied with
     <br/>
     molecular simulation methods.”
    </em>
   </p>
  </blockquote>
  <p>
   Now, in this fragment, we can have a glance of the complexity of living systems, with the mention of the 3D structure of molecules inside cells as explanatory of the functions the perform, although the modelling of these structures is very demanding in terms of computation resources.
  </p>
  <hr/>
  <script src="http://cdnjs.cloudflare.com/ajax/libs/highlight.js/8.1/highlight.min.js">
  </script>
  <script>
   hljs.initHighlightingOnLoad();
  </script>
  <script src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML" type="text/javascript">
  </script>
  <script type="text/javascript">
   MathJax.Hub.Config({"showProcessingMessages" : false,"messageStyle" : "none","tex2jax": { inlineMath: [ [ "$", "$" ] ] }});
  </script>
 </body>
</html>